Temporal vector feature handling fix

rduivenvoorde opened the pull request:
@nyalldawson this is the first part of work I did after A LOT of thinking about the practical use if this (in my daily work).
I am aware that I change things which other (you) have been thinking about also a lot probably.
I also consulted @anitagraser just to be sure I am on the right track.

This is all about this: https://github.com/qgis/QGIS/issues/38468
which I at that moment fixed partly in https://github.com/rduivenvoorde/QGIS/commit/336936ba2401275e0ec613c3992e9b8700338b9d by NEVER taking the timestep limits into account.

I think that is the right thing to do, but I only implemented that in case of 'ModeFeatureDateTimeInstantFromField' while in that QgsVectorLayerTemporalProperties::createFilterString method there are actually implementations for the other (around 4) possible modes....
This WIP is about fixing only one of them now, and discuss this, before doing the others too (and writing tests for them).

I now also take into account that people choose a mode: '2 fields for end and start' BUT... then only define one of them.
Actually (after writing all the testing) I stand corrected, and I think we should just not allow that (and throw an exception or so).
Note that this is NOT about when DATA has a null value (which is handled by the OR "xxx_field" IS NULL part of the expression.

Argh... thinking aloud now... Now I fully ignore the fact that you can programmatically define that you should (or not) take the limitis into account... While probably a better solution, would be... to DEFAULT to a range NOT taking BOTH limits into account (while now we take only the start limit into account)....
Well at least I now have written a lot of testing code.

To recap:
- the timestep frame is ALWAYS a timeFRAME
- when really testing an event with a timeframe (start AND end): by default do NOT take limits into account
- when testing an event with a zero-sized timeframe (a MOMENT like a tweet), then DO take limits into account
- we should not be able to have a null in one of the field params (we CAN have it in data)

Not making it easier for others to follow I'm afraid.... :-(

rduivenvoorde committed:
Fix (only) for ModeFeatureDateTimeStartAndEndFromFields

When the feature event is a moment (start-end), then you should
use start (including) AND end (NOT included).
But when the feature event has a range, you should NEVER take
the timeframe limits into account (as then you will have double
hits in case of testing on the exact timeframes)

A test to check all kind of timeframes, null data etc is added.

This does NOT yet fix the old tests.

rduivenvoorde committed:
Fix other tests

nyalldawson commented:
> I now also take into account that people choose a mode: '2 fields for end and start' BUT... then only define one of them.
Actually (after writing all the testing) I stand corrected, and I think we should just not allow that (and throw an exception or so).

I disagree with this -- to me that's removing a possibly valid use case, where data "starts" at some field and never ends (or "ends" at some field but was always beginning!). Why would we want to prevent this?

rduivenvoorde commented:
> > I now also take into account that people choose a mode: '2 fields for end and start' BUT... then only define one of them.
> > Actually (after writing all the testing) I stand corrected, and I think we should just not allow that (and throw an exception or so).
> 
> I disagree with this -- to me that's removing a possibly valid use case, where data "starts" at some field and never ends (or "ends" at some field but was always beginning!). Why would we want to prevent this?

Mmm, at first I thought we were mixing up things (as in mixing up defining the FIELDS vs possible DATA nulls), but on second thought, you are probably right... 

That one use-case cannot be defined in one of the other 'modes'?

Then my proposed code is working, except(!) it does not handle Ranges with non standard (included begin-limit, NON-inluded end limit)... 
I have to think about that.... I would be happy to keep your/the old code, except it does not work in my (test) case...

But do you agree on the fact that the test I added currently is behaving OK? This is the big thing here: that in the old implementation an event with an equal length of a timeframe and starting on the timeframe start could popup in 2 timeframes (which IS theoretically a possible scenario, but in practice not very intuitive, I think)...

(In Geoserver I read somewhere that they always substract a microsecond from one of the ranges to fix this issue...)

nyalldawson commented:
> That one use-case cannot be defined in one of the other 'modes'?

Not at the moment -- maybe we should add two new explicit modes though to handle this though -- "start time from field"/"end time from field" (but we'd also need to handle upgrading existing projects with only one field set to the corresponding new mode)

rduivenvoorde commented:
Another issue with current/my code is that it does not handle correctly the case where the VALUES of the actual startfield and endfield are actually identical, with this defining a moment-event instead of an event with a duration >0.
Which I think cannot even be handled with 1 expression, as this case can only be tested for every record and IF start-value==end-value you actually want the expression to behave differently...

So maybe adding 2 modes could help, but I find the possibilities already becoming pretty complex...

rduivenvoorde committed:
Proper (more intuitive) handling of temporal vector data

Crux is:
- when both sliding view-frame (always has some duration) AND the data
has an duration, then you should NOT take the limits of the view-frame
into account (else a feature would show up in 2 view frames IF it is on
the limits/border of those frames)
- when the data is momentarily (having NO duration, so just one moment),
you SHOULD take at least one limit into account, often the beginning)

rduivenvoorde commented:
Ok, I reverted part of my own work (because I think Nyall's work was actually better)

Main change now is in this part:

https://github.com/qgis/QGIS/pull/40989/files#diff-b4a4c7eb1b9251a97a6e0bdaad9d253fcedd249631167e366cd3f4c27f5789f6R108-R110

So: the default QgsDateTimeRange created should NOT take the limits into account.
ONLY when the data has NO duration the beginning limit (of the view frame) should be taken into account.

I think we still have small issue (which we cannot fix yet), is if when we define data/a layer with a start- and end-field, and the VALUE of those are identical (actually meaning the event does not have a duration)... But we cannot fix this (I think) because we use one filter string for the full data set. But I think this is a corner case which hopefully will not bite somebody.

Another thing which I though of: we actually conceptually glued the timestep and the timeframe of the filter together: if you define your step to be '1 hour', you also step through your data in steps of one hour. While off course it is perfectly ok to have a timestep (filter window) of 1 hour, but step through your data in steps of 30 minutes. (I understand this 'feature' is less usable in WMS-T and mesh temporal data, so let's make it a future feature request).
To make it more complex: I can also imagine a view frame with a duration of zero: so actually a thin slice through the data (maybe can be seen as a tiny range).

I added a silly but very practical set of tests with a lot of comments, I hope people will have time to build this and check this new filters with their own vector data. IF you think I made a mistake (either conceptually, or because this code does not fit your data, please let me know or discuss it below).  
IF other people agree on current filter rules (for vector data), I think it is worth backporting this to 3.16
BUT only if people fully agree on current code implementation.

I ran flake8 and ./scripts/prepare_commit.sh, not sure why code layout is still failing?

rduivenvoorde commented:
Mmm, still not okay...

This logic substract the time of event from the start, I do not understand?
I have 2 datasets, one is working with this, and the other is not... have to investigate...

```
          return QStringLiteral( "(%1 %2 %3 AND %1 %4 %5) OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
               dateTimeExpressionLiteral( range.begin() + QgsInterval( -mFixedDuration, mDurationUnit ) ),
               range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
               dateTimeExpressionLiteral( range.end() ) );
```

Did the used 'range' maybe change? That before, it did not contain an actual range of size "mFixedDuration, mDurationUnit"?
Now having 1 hour data and 1 hour step it becomes: 
"(\"Time\" > make_datetime(2019,9,1,6,0,0) AND \"Time\" < make_datetime(2019,9,1,8,0,0)) OR \"Time\" IS NULL"

While it looks like this should now be just the same as in the logic for features with start and end-times defined?:
"(\"Time\" > make_datetime(2019,9,1,7,0,0) AND \"Time\" < make_datetime(2019,9,1,8,0,0)) OR \"Time\" IS NULL"

anitagraser commented:
> Another thing which I though of: we actually conceptually glued the timestep and the timeframe of the filter together: if you define your step to be '1 hour', you also step through your data in steps of one hour. While off course it is perfectly ok to have a timestep (filter window) of 1 hour, but step through your data in steps of 30 minutes.

+1 we had that feature request on the TimeManager tracker for a long time but never got around to implementing it

anitagraser commented:
@rduivenvoorde can you provide a video of the now implemented behavior (similar to what you posted in https://github.com/qgis/QGIS/issues/38468)? (I don't have a build environment / time to set it up right now) and having trouble imagining the implications of 

> the logic used to determine the overlap between to ranges should: either use BOTH limits (not very practical: as it make features show up in two 'classes') or NONE limits (is what we do now).

Does it mean: if a feature has a time range from 03:00-04:00 and I have a frame from 03:00-05:00 it will not show at all?

rduivenvoorde commented:
> Does it mean: if a feature has a time range from 03:00-04:00 and I have a frame from 03:00-05:00 it will not show at all?

No, that will 'just show' as these frames clearly overlap. But if you have a feature with 3-4 and one with 4-5, and your step/view frame is 3-4, then ONLY the first one will be shown.  I'll try to make a video as you propose.

Samweli commented:
Hi @rduivenvoorde 

> While probably a better solution, would be... to DEFAULT to a range NOT taking BOTH limits into account (while now we take only the start limit into account)....

I'm -1 on this, I think we should keep the range as it is (taking into account only the start limit) and deal with challenges which the different temporal cases that vector layers provide(eg  a layer having a field with timestamp instant) when creating filtering strings for the layers. It makes more sense to me for the QgsTemporalNavigationObject to provide a range which has start limit taken into account than NOT taking into account any of the limits.

Samweli commented:
Hi @rduivenvoorde 

> While probably a better solution, would be... to DEFAULT to a range NOT taking BOTH limits into account (while now we take only the start limit into account)....

I'm -1 on this, I think we should keep the range as it is (taking into account only the start limit) and deal with challenges which the different temporal cases that vector layers provide(eg  a layer having a field with timestamp instant) when creating filtering strings for the layers. It makes more sense to me for the QgsTemporalNavigationObject to provide a range which has start limit taken into account than NOT taking into account any of the limits.

rduivenvoorde commented:
@Samweli maybe we can have a chat about this? I really did a lot of thinking and testing. 
I think that working with 2 ranges/frames (so when both the filter-view is a range/frame/period AND the data consists of frames) it is plain wrong to use the limits. It only makes sense when you data is 'point in time' based (so no duration) (and most classifications are about point/scalar data's so THERE using at least one limit makes sense.
This is not depending on type of data, take for example, hourly measurements: starting at every hour:00 
Now if you filter those with a filter also of 1 hour, AND you take the begin into account, then you can NEVER have just 1 result, not even if you follow the hours exactly.

Current implementation makes the Temporal Controller unusable to me.
So let's have a chat, maybe you can show me some data where it is important to have the start limit included? 

I googled this earlier: 
https://wiki.c2.com/?TestIfDateRangesOverlap

roya0045 commented:
I agree with Richard here. Having support for common temporal data queries [such as these usage with mssql](https://docs.microsoft.com/en-us/sql/relational-databases/tables/temporal-table-usage-scenarios?view=sql-server-ver15) should not be excluded, Oracle also has similar data structure and so does Postgres with an extension.

Samweli commented:
> @Samweli maybe we can have a chat about this

yes we can @rduivenvoorde

rduivenvoorde committed:
Handle comment about NOT changing the TemporalNavigationObject

To make the temporal controller work for vectordata we override/change(!) the
actual range object in the 'createFilterString' method.
One of the results/drawbacks of this is that the (theoretical) use of passing it a
QgsDateTimeRange with useBegin and useEnd is actually not possible.
The positive result is that we are not changing the TemporalNavigationObject
so not breaking the use of the in other providers...

rduivenvoorde commented:
@Samweli I reworked this to NOT change the TemporalNavigationObject, but ONLY change the vector provider part. Plz have a look.

Samweli commented:
> @Samweli I reworked this to NOT change the TemporalNavigationObject, but ONLY change the vector provider part. Plz have a look.

Ok @rduivenvoorde, I ll get back to you...

Samweli commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -321,32 +327,165 @@ def testExpressionMode(self):
+        for fv in [fv4, fv5, fv6]:
+            f = QgsFeature()
+            f.setAttributes(fv[2:])
+            f.setGeometry(QgsGeometry.fromPointXY(QgsPointXY(fv[0], fv[0])))
```
`f.setGeometry(QgsGeometry.fromPointXY(QgsPointXY(fv[0], fv[1])))`

Samweli commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -321,32 +327,165 @@ def testExpressionMode(self):
+        self.assertEqual(len(result), 0)
+
+        frame = QgsDateTimeRange(QDateTime(QDate(2020, 8, 26), QTime(2, 0, 0)), QDateTime(QDate(2020, 8, 26), QTime(3, 0, 0)), False, False)
+        # timestep frame on exact one data range frame: 2:00-3:00
```
what if frame range is from 1:00 - 4:00, should we expect results with 3 features or 2? maybe we need to add it as another test case at least I think it is an important one.

rduivenvoorde replied:
IF having features with a duration of one hour, I would expect 3: those from 1-2, 2-3 and 3-4
I'll add a test case

rduivenvoorde commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -427,24 +433,27 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
-               dateTimeExpressionLiteral( range.end() ) );
+               ">=",  // always taking the begin into account
+               dateTimeExpressionLiteral( vectorRange.begin() ),
+               vectorRange.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```
Yes, I agree:

- when you have a view range of length > 0 and data points (without a duration), and you want you points to fall always in at least one range class, you really have to always include one range limit and exclude the other.
Without this you will loose data.

In my view the whole excercise to be able to include or exclude limits for QgsDateTimeRange is theoretical and NOT practical.

So in this case I would even be happy to create something here like: `QStringLiteral( "(%1 >= %3 AND %1 < %5)` if you think that is more clear.
And while `QStringLiteral( "(%1 > %3 AND %1 <= %5)` (so including end only) is also a theoretical option, I doubt user will want this...

rduivenvoorde commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -427,24 +433,27 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
-               dateTimeExpressionLiteral( range.end() ) );
+        // Working with features with events with a duration
+        return QStringLiteral( "(%1 %2 %3 AND %1 %4 %5) OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
+               vectorRange.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
Yep, and as above, here the story is the other way around:
- having both a view range with a duration AND features with a time and a duration you NEVER want to add limits into your filter range, because that would mean that your features will fall into 2 range classes ( see https://github.com/qgis/QGIS/issues/38468 )

rduivenvoorde commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -511,23 +520,23 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
                QgsExpression::quotedColumnRef( mEndFieldName ),
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
-               dateTimeExpressionLiteral( range.begin() ) );
+               vectorRange.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
See above.... if anybody makes a valid case in which you want you want a feature to show up in two classes (time periods), I will rest my case. But untill that I think these are all these options are theoretical (and so possible), but not practical in use case.

Samweli replied:
The issue of feature showing up in two classes was happening when the layer has one datetime field, is it also a case when layer has two datetime fields(start_field and end_field)? If so then we definitely need to amend the logic that deals with two datetime fields.

rduivenvoorde replied:
@Samweli yes, IF you take (whatever) limit (one OR more), then you will end up with this problem.
THAT is exactly what I'm trying to fix here...

```
-----0_____x-----0-----0-----0 possible range/filter steps '0' = NOT taking limit into accoutn, 'x' is DO take limit into account
-----|-----|-----|-----|-----|   data
--------X-----X---------------   true in filter

-----0_____0-----0-----0-----0 possible range/filter steps '0' = NOT taking limit into accoutn, 'x' is DO take limit into account
-----|-----|-----|-----|-----|   data
--------X---------------------   true in filter

Another option/route (taken by the Geoserver people if I'm correct), is to substract a millisecond from the data OR ranges to handle this.... But let's not make it even more complex

```

rduivenvoorde commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -165,18 +169,20 @@ def testSingleFieldMode(self):
         props.setFixedDuration(3)
         props.setDurationUnits(QgsUnitTypes.TemporalDays)
-        range = QgsDateTimeRange(QDateTime(QDate(2019, 3, 4), QTime(11, 12, 13)), QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10)))
+        range = QgsDateTimeRange(QDateTime(QDate(2019, 3, 4), QTime(11, 12, 13)), QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10)), includeBeginning=False, includeEnd=False)
```
I know, I was just making the tests work again here I think...

rduivenvoorde commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -208,7 +214,7 @@ def testDualFieldMode(self):
         context.setLayer(layer)
 
-        range = QgsDateTimeRange(QDateTime(QDate(2019, 3, 4), QTime(11, 12, 13)), QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10)))
+        range = QgsDateTimeRange(QDateTime(QDate(2019, 3, 4), QTime(11, 12, 13)), QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10)), includeBeginning=False, includeEnd=False)
```
idem. To me it would be ok to remove this test cases, but untill my PR is merged or somebody agrees with my reasoning I thought not to remove those test cases.

rduivenvoorde commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -321,32 +327,165 @@ def testExpressionMode(self):
+        for fv in [fv1, fv2, fv3]:
+            f = QgsFeature()
+            f.setAttributes(fv[2:])
+            f.setGeometry(QgsGeometry.fromPointXY(QgsPointXY(fv[0], fv[0])))
```
Ah, good catch thanks! That is funny that I still get my tests done...

rduivenvoorde committed:
Fix wrong xy in test

rduivenvoorde committed:
Add a test as requested

timlinux commented:
We had a long discussion about the expected behaviour and these diagrams summarise the behaviour that this PR introduces:

The current behaviour:

![image](https://user-images.githubusercontent.com/178003/107399261-68588f00-6af8-11eb-96ad-a12098846765.png)


Proposed behaviour:


![image](https://user-images.githubusercontent.com/178003/107399137-4d861a80-6af8-11eb-849a-530a67749e94.png)


![image](https://user-images.githubusercontent.com/178003/107399100-4101c200-6af8-11eb-99b8-c4b474153c75.png)

timlinux commented:
Here is the draw.io document accompanying above

[QGIS PR 40989 Diagram.zip](https://github.com/qgis/QGIS/files/5952942/QGIS.PR.40989.Diagram.zip)

timlinux commented:
Just a comment from an end user, I built @rduivenvoorde 's branch and it now behaves as expected as per the diagrams above. +1 from me to merge if the code is up to scratch!

rduivenvoorde commented:
@nyalldawson can you please have a look at this? Sorry for the long conversations etc.

rduivenvoorde commented:
@nyalldawson that is a new one for me :-) 'TRICKY!!' ....

What to you find tricky on this?

nyalldawson commented:
> What to you find tricky on this?

I have to read a whole essay in order to review it :rofl:

nyalldawson commented:
![image](https://user-images.githubusercontent.com/1829991/109237226-e026f900-781c-11eb-97e5-db7d9f75c048.png)

While I agree with the intention of this PR and the need to avoid duplicate point visibility, I disagree with the last illustration in this diagram. If you have a map set to a time range of 10:00 - 11:00, and a point feature with a datetime stamp of 11:00, then you'd most definitely expect to see that point visible (it's a bug if not!).

I think there's a distinction here between the time handling running in an animation mode vs running in the time filter mode. In time filter mode the point should show for the range of 10:00-11:00, and the user has the freedom to change the upper range to 10:00-10:59:59 if they don't want to see this. In animation mode I think you're correct in saying that the time range shouldn't include the upper limit, so that there's no overlap between consecutive frames.

anitagraser commented:
> I think there's a distinction here between the time handling running in an animation mode vs running in the time filter mode.

I think it is a bad idea to have two different behaviors in animation and filter mode. Users will be very confused.

rduivenvoorde commented:
> While I agree with the intention of this PR and the need to avoid duplicate point visibility, I disagree with the last illustration in this diagram. If you have a map set to a time range of 10:00 - 11:00, and a point feature with a datetime stamp of 11:00, then you'd most definitely expect to see that point visible (it's a bug if not!).

I do not agree: just to be sure: if you do what you propose, and stepping in steps of exactly one hour, you will see that point BOTH in 
10:00-11:00 AND in 11:00-12:00 making your data fall in 2  classes, something (I think) you never want when you are classifying your data (what this actually is if I'm correct)

> I think there's a distinction here between the time handling running in an animation mode vs running in the time filter mode. In time filter mode the point should show for the range of 10:00-11:00, and the user has the freedom to change the upper range to 10:00-10:59:59 if they don't want to see this. In animation mode I think you're correct in saying that the time range shouldn't include the upper limit, so that there's no overlap between consecutive frames.

But that is another issue with current Time Controller: the 'Size of the Timesteps' you step through your data is always the same as the 'Size of the Viewing Time Frames' (event duration)... Which in current implementation makes that IF you for example configure the layer with a Start Field and an event duration of 59minutes:59seconds you will also step with (the rather odd) steps of 59minutes:59seconds through your data.

I agree with @anitagraser there should be no distinction between animation vs running mode.

My conclusion: whatever you are classifying: ranges or timestamps you will never want your data to show up in 2 classes. And that is exactly what this PR is for :-)

rduivenvoorde commented:
> My point: whatever you are classifying: ranges or timestamps you will never want your data to show up in 2 classes. And that is exactly what this PR is for :-)

I now realize myself that with a gliding 'view/filter', data with ranges will always fall in two classes anyway, so I'm wrong in this.
It's just: I see/feel something is not behaving right, but I cannot make it clear apparently. I'm fine with closing this PR.
But for my initial use-cases there is an border issue in current implementation:

- having hourly data of exact one hour (xx:00-xx:00), stepping through it in exact 1 hour frame, I think there should be a way to just see one hour-data (per frame). 
My practical example: model output generated per hour per square km (so with start and end, OR with start and 1 hour 'duration'), you want to be able to step through it, seeing 1 result per step)?

- having hourly events (moments), stepping through it in exact 1 hour timesteps, I think they should not fall in two (temporal) categories.
My practical example: measuring posts which generate hourly (integrated) data points, if you step through it per hour, you do not want to see posts appear in two timesteps, just because some lay on the border of the timeframe/step?

A total other way to make the above two (my) use cases possible, is to split the 'filter/duration-frame' and the 'step-frame': then I would be able to filter on 0:59:59.999 frames, but step through in 1:00:00.000 steps.
OR 
make it possible to truly define the use of begin and end in the QgsDateTimeRange used (either in the data or in the steps)

I'm very sorry to look so determined, but either I 'just do not see/understand' my issue, or I make a giant thought error which apparently others just do not have/see.

anitagraser commented:
> * having hourly events (moments), stepping through it in exact 1 hour timesteps, I think they should not fall in two (temporal) categories.

💯

nyalldawson commented:
@rduivenvoorde 
> I do not agree: just to be sure: if you do what you propose, and stepping in steps of exactly one hour, you will see that point BOTH in
10:00-11:00 AND in 11:00-12:00 making your data fall in 2 classes, something (I think) you never want when you are classifying your data (what this actually is if I'm correct)


Right, but look at the below screenshot -- I would 100% expect the 11:00 point to be visible here, and anything else would be misleading and a bug:

![image](https://user-images.githubusercontent.com/1829991/109442376-c2a79880-7a83-11eb-85b9-c930d7302ef7.png)

That's in the temporal filter mode. 

Now if I flick to animation mode:


![image](https://user-images.githubusercontent.com/1829991/109442478-19ad6d80-7a84-11eb-8f67-9f43e4f9959d.png)

This looks plain wrong to me. The 11:00am point SHOULD be visible given that the UI reports this as the time range:

![image](https://user-images.githubusercontent.com/1829991/109442514-2cc03d80-7a84-11eb-851c-2e01994284ef.png)

If we are going to make the beginning or end of the range excluded then we should rework these widgets to make that obvious to users, ie. 

"Range >= 2021-03-01 10:00:00 to < 2021-03-01 11:00:00"

rduivenvoorde commented:
> Right, but look at the below screenshot -- I would 100% expect the 11:00 point to be visible here, and anything else would be misleading and a bug:

Well, in my experience you do NOT want to see 11:00 if you are filtering on 10:00 to 11:00. It's the same as with styling rules, silently we expect to INCLUDE the start, but EXCLUDE the end limit: 10:59:59.999 is to included though...

And indeed to make that super clear we could work with >= and < signs, as you propose:
"Range >= 2021-03-01 10:00:00 to < 2021-03-01 11:00:00"

Setting those limits (as is possible in code) is actually super complex if you start to have mixed layer types:
- if you have 'point in time' data you can work with the different options
- if you have data with a duration (frames) it is either take BOTH or take NEITHER 
That is what I described earlier with 'theoretical' options (as in: possible in code) vs logical/practical options...

In my experience people tend to agree on: always include the lower limit and exclude the upper limit.

neteler commented:
> Well, in my experience you do NOT want to see 11:00 if you are filtering on 10:00 to 11:00. It's the same as with styling rules, silently we expect to INCLUDE the start, but EXCLUDE the end limit: 10:59:59.999 is to included though...

FWIW, in GRASS GIS, specifically TGRASS, time intervals are defined as follows:

- "the time interval is a left closed right open interval".

_Source:_ p.5 in 
- Gebbert, S., Pebesma, E. 2014. TGRASS: _A temporal GIS for field based environmental modeling_. Environmental Modelling & Software 53, 1-12 ([DOI](http://dx.doi.org/10.1016/j.envsoft.2013.11.001); preprint: http://pebesma.staff.ifgi.de/tgrass.pdf)

which is AFAIK conformant to the implementation in R (also, given the authors of the paper).

Practical examples (also further article references) are here: https://grasswiki.osgeo.org/wiki/Temporal_data_processing

gioman commented:
> which is AFAIK conformant to the implementation in R (also, given the authors of the paper).

I think there is a compelling argument to support/follow what R and others do.

rduivenvoorde commented:
> FWIW, in GRASS GIS, specifically TGRASS, time intervals are defined as follows:
> 
>     * "the time interval is a left closed right open interval".

Thanks @neteler I (tried to) read the article, though I really need some 'translation', to map it to the QGIS TimeManager/Temporal Controller... :-)

We do not so much temporal relations (as in the article), do we? But more 'temporal classification' as in: show me from my data (all layers) all features that are within this time interval. For data with 'Time Instances' (as it is called in the article) it is rather clear (to me): you slice with 'including start, excluding end' ("the time interval is a left closed right open interval").

But when you are going to filter 'Time Interval' data (either a start/end as attributes, or a start attribute plus a duration) with a given Time Interval Filter (our TimeFrame: our 'filter' is always an interval), from what I read you either have to add BOTH or NO of the limits... of which I concluded to use NO limits.

veroandreo commented:
> > Well, in my experience you do NOT want to see 11:00 if you are filtering on 10:00 to 11:00. It's the same as with styling rules, silently we expect to INCLUDE the start, but EXCLUDE the end limit: 10:59:59.999 is to included though...
> 
> FWIW, in GRASS GIS, specifically TGRASS, time intervals are defined as follows:
> 
>     * "the time interval is a left closed right open interval".

Precisely! In TGRASS we define intervals with a start and end time, where the end time is not included in the current interval but it becomes the start time of the immediate next one, if intervals are regular. 

Furthermore, we have what we call time instances that are defined only with a start time (and `None` for end time), i.e., it's something instantaneous or so. Then with temporal sampling and temporal topology we can relate different time series. 

So, the point of the above example with a time stamp 11:00:00 (in TGRASS this is a time instance), if we compare it with one hour intervals of another time series as defined in TGRASS (i.e., 9:00:00-10:00:00, 10:00:00-11:00:00, 11:00:00-12:00:00), would fall in the last interval.

rduivenvoorde commented:
@veroandreo Thanks! 

From https://grass.osgeo.org/grass79/manuals/t.vect.algebra.html I conclude that what we are doing is 'temporal' selection, in which our 'Temporal Controller'-Frame is a Time-Interval and our data is either a Time-Interval or a Time-Instance.

The crux of this PR is that I think we have to handle these differently:
To check if two an 'time-interval' and a 'time-instance' 'overlap', you can work with a "left closed right open interval"
But if you have two 'time-interval's in *our* code [0] in which we create a 'filter'-string you should use "left open right open intervals", is my idea?

I've tried to create an issue with testdata and testprojects to show what this PR is actually fixing:
https://github.com/qgis/QGIS/issues/41970
Plz have a look and let me know where I'm wrong.

[0] https://github.com/qgis/QGIS/blob/master/src/core/vector/qgsvectorlayertemporalproperties.cpp#L414-L563

veroandreo commented:
> @veroandreo Thanks!
> 
> From https://grass.osgeo.org/grass79/manuals/t.vect.algebra.html I conclude that what we are doing is 'temporal' selection, in which our 'Temporal Controller'-Frame is a Time-Interval and our data is either a Time-Interval or a Time-Instance.
> 
> The crux of this PR is that I think we have to handle these differently:
> To check if two an 'time-interval' and a 'time-instance' 'overlap', you can work with a "left closed right open interval"
> But if you have two 'time-interval's in _our_ code [0] in which we create a 'filter'-string you should use "left open right open intervals", is my idea?

IMHO, a time interval should be consistently defined throughout the software, i.e., always in the same way. It might be confusing and difficult to maintain afterwards if the definition/behaviour differs here and there. 

For the case of 2 time series with time intervals (i.e., maps within them have time interval timestamps), you have many different possible temporal relationships according to which time series is your focus. In GRASS, the temporal relationships defined are [here](https://grass.osgeo.org/grass79/manuals/t.vect.algebra.html#temporal-relations): 

![image](https://user-images.githubusercontent.com/20075188/109667287-48087700-7b70-11eb-9483-b87af423383b.png)

In any case a time interval is always defined as "left closed right open", otherwise these relationships would become inconsistent. Maybe, it's obvious for you, but I do not understand why you would need to change interval definition for the case of two time intervals. 
 
> I've tried to create an issue with testdata and testprojects to show what this PR is actually fixing:
> #41970
> Plz have a look and let me know where I'm wrong.
> 
> [0] https://github.com/qgis/QGIS/blob/master/src/core/vector/qgsvectorlayertemporalproperties.cpp#L414-L563

I'm not familiar with QGIS code TBH, I can only say how it works in GRASS and what I know from having used TGRASS quite  frequently.

nyalldawson commented:
@rduivenvoorde ok, I'm willing to put aside my concerns on the condition that you update the UI to explicitly reflect the range inclusion / exclusions, so that there's zero chance for misinterpretation by users. (I'm referring to the temporal dock widget and the various layer temporal properties panels too)

rduivenvoorde commented:
@nyalldawson I think it is just not so easy, I do not think I'm able to do that

I think temporal filtering is more complex then our current setup is handling:
- the range-inclusions are provider specific, so cannot be made more clear in the controller (as we 'mix' temporal props/capabilities in your project)
- filtering (handling range limit inclusions) of instant data or interval data should be handled differently (while we force (in the ui) one model)

rduivenvoorde commented:
I would think this 'fixed' showing of limits would not be accepted (as earlier pr's from me to make this explicit in code (not in ui)) was not accepted... 

Not sure if people now can try to offer TimeRanges (as filter) tot the controller with different limits.
BUT... as said: I think this should actually not possilbe (the fact that you can offer a range with both closed or both open limits, or open start and closed end)...

rduivenvoorde committed:
Fixing the use of the temporal limits for vector features

In practice you do not want the filter range to be able to determine
the fact if limits of data ranges should be in- or excluded.

rduivenvoorde committed:
Add minimal text about inclusive/exclusive limits

rduivenvoorde commented:
ping @nyalldawson 
Is it easiest if I rebase for this conflict? Or should I just take over the master version lines?

nyalldawson commented:
@rduivenvoorde 
> Or should I just take over the master version lines?

You should copy the master version lines over yours

rduivenvoorde committed:
Merge branch 'master' into TemporalFeatureFix

rduivenvoorde commented:
Mmm, hope that went ok. I used the 'github web editor' to resolve the conflict ...

rduivenvoorde commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
-             range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
              dateTimeExpressionLiteral( range.begin() ),
              QgsExpression::quotedColumnRef( mDurationFieldName ) );
-      break;
```
Qt creator tells me this break will never be reached, so I removed it

rduivenvoorde commented:
@nyalldawson anything else missing here? Ui stuff I'm discussion in #42548

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
           return QString();
       }
-      return QStringLiteral( "(%1 %2 %3 OR %1 IS NULL) AND ((%1 + %4 %5 %6) OR %7 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
-             range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```
we need to leave the `range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),`  logic here -- see above

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -443,22 +443,20 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
     {
       if ( mAccumulateFeatures )
       {
-        return QStringLiteral( "(%1 %2 %3) OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
```
we need to leave the range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ), logic here -- while the gui isn't creating ranges which includes the end, the raw api may still be used to set a map range with inclusive end and it's important it's correctly handled here

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -443,22 +443,20 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
+        // You really need to take at least ONE limit into account (in practice mostly the beginning)
+        return QStringLiteral( "(%1 >= %2 AND %1 < %3) OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
                dateTimeExpressionLiteral( range.begin() ),
-               range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```
we need to leave the range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ), logic here -- see above

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -443,22 +443,20 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       else if ( qgsDoubleNear( mFixedDuration, 0.0 ) )
       {
-        return QStringLiteral( "(%1 %2 %3 AND %1 %4 %5) OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
I agree with the the change to the beginning of range logic :+1:

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
+      return QStringLiteral( "(%1 < %2 OR %1 IS NULL) AND ((%1 + %3 > %4) OR %5 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
              dateTimeExpressionLiteral( range.end() ),
              intervalExpression,
-             range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
I agree with the the change to the beginning of range logic :+1:

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
+        return QStringLiteral( "(%1 < %2 OR %1 IS NULL) AND (%3 > %4 OR %3 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
                dateTimeExpressionLiteral( range.end() ),
                QgsExpression::quotedColumnRef( mEndFieldName ),
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
I agree with the the change to the beginning of range logic :+1:

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       if ( !mStartFieldName.isEmpty() && !mEndFieldName.isEmpty() )
       {
-        return QStringLiteral( "(%1 %2 %3 OR %1 IS NULL) AND (%4 %5 %6 OR %4 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
-               range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```
we need to leave the range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ), logic here -- see above

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       {
-        return QStringLiteral( "%1 %2 %3 OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mEndFieldName ),
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
+        return QStringLiteral( "%1 > %2 OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mEndFieldName ),
```
I agree with this logic :+1:

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -514,37 +512,31 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       {
-        return QStringLiteral( "%1 %2 %3 OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
-               range.includeBeginning() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
+        return QStringLiteral( "%1 < %2 OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
```
I think the logic here is broken (completely in the current code!)

I believe the correct one is:

```
return QStringLiteral( "%1 %2 %3 OR %1 IS NULL" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
               range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -554,23 +546,19 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
+        return QStringLiteral( "((%1) < %2) AND ((%3) > %4)" ).arg( mStartExpression,
                dateTimeExpressionLiteral( range.end() ),
                mEndExpression,
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
```
I agree with the the change to the beginning of range logic +1

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -554,23 +546,19 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       if ( !mStartExpression.isEmpty() && !mEndExpression.isEmpty() )
       {
-        return QStringLiteral( "((%1) %2 %3) AND ((%4) %5 %6)" ).arg( mStartExpression,
-               range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
```
we need to leave the range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ), logic here -- see above

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -554,23 +546,19 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       {
-        return QStringLiteral( "(%1) %2 %3" ).arg( mEndExpression,
-               range.includeBeginning() ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
+        return QStringLiteral( "(%1) > %2" ).arg( mEndExpression,
```
I agree with this logic :+1:

nyalldawson commented on src/core/vector/qgsvectorlayertemporalproperties.cpp:
```diff
@@ -554,23 +546,19 @@ QString QgsVectorLayerTemporalProperties::createFilterString( const QgsVectorLay
       }
       else if ( !mStartExpression.isEmpty() )
       {
-        return QStringLiteral( "(%1) %2 %3" ).arg( mStartExpression,
```
I think the logic here is broken (completely in the current code!)

I believe the correct one is:

```
return QStringLiteral( "(%1) %2 %3" ).arg( mStartExpression,
    range.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
    dateTimeExpressionLiteral( range.end() ) );
```

nyalldawson commented:
Almost there! I agree with the majority of the changes, but we need to rollback the bits which ignore whether the map's ranges include the end of the range. While that situation is not possible to encounter using the temporal controller GUI, it IS possible to encounter is someone is using the raw api to render a map and they deliberately set the end of the map's temporal range to be inclusive. In that case we need to ensure it's correctly handled in this class.

rduivenvoorde commented:
Mmm, because of the Merge of master I'm a little lost about how to handle your comment(s).

I think I will try to 'roll back' to the version in which I do use the 'vectorRange' (a range with both excluding limits). But only *use* this vectorRange when the 'range'-argument is default (== including begin, excluding end).  
I do not see a way of handling both moment-events vs range-events in one function.

rduivenvoorde commented:
Going to close this one.

On hindsite, I think my 'solution' here hacks around some conceptual issues in the current use of data with ranges.

I think it is better to fix/discuss this conceptual issues first, hopefully when somebody with some (financial) interest also hit my issue(s).

CoryAlbrecht commented:
The discussion here seems to assume that all temporal map data can have have duration lengths which are smaller than the step size, but that isn't the case. 

For example, if you have a thermometer that you record at `YYYY/MM/DD 00:00:00.000` every day and show that on the map as a filled circle with the colour representing the temperature, you'll be wanting to step through at 1 day per step, (not 1 millisecond) and you only want to show `YYYY/MM/DD 00:00:00.000 → YYYY/MM/DD 23:59:59.999` and not `YYYY/MM/DD 00:00:00.000 → YYYY/MM/DD+1 00:00:00.000` because that would mean there'd be two two coloured circles on the map.

With events like that, the `$X ≤ t < $Y` logic works because you're not really wanting `$timestart + $stepsize` but rather `$timeStart + $stepSize - $smallestTimeResolutionUnitPossible` because your data have the concept of durations that can be shorter than your frame step size.

But when your minimum duration can't get shorter than your step size, that isn't the case.

If somebody tells you a conference goes from August 1st to the 7th, it's a seven day long conference. However the logic used in the QGIS temporal controller is a bit like saying "7 - 1 = 6 therefore the conference is only 6 days long", which is incorrect.

Consider a big conference centre with multiple halls of varying size that are rented out to conventions for one or more whole days at a time. The centre's website has a map which you can step through day by day with empty halls showing in grey and used halls showing as unique colours. One conference (assigned green) books hall A, the smallest one, for August 1st through 7th, and a second conference (assigned blue) books the same hall from August 8th through the 14th. With the current QGIS logic, if you went looking to book the small hall stepping through the first half of August, when you get to the 7th it will show hall A as grey leading you to think that hall A is available on that day. But it is not.

rduivenvoorde commented:
Hi @CoryAlbrecht , please note that this issue is already closed (and implemented).

I agree there will proably some usecases not yet covered, but did you try some of your cases? Because (knowing the internals of the timeframe calculation: always in (micro?)seconds) I think your conference example should just work: it's not that we substract one used time unit from the set. In this case we do 7 days * secondsInADay, and that is the duration.

I think current implementation is not yet ideal, but works for most cases. If you have a case for which it does NOT work, please create a separate issue, plus add some data+projectfile to it, so we can have a look at it. 
Thanks

gioman commented:
> . If you have a case for which it does NOT work, please create a separate issue, plus add some data+projectfile to it, so we can have a look at it.

see here https://github.com/qgis/QGIS/issues/44589

CoryAlbrecht commented:
OK, then how do you recommend I work around this issue in the meantime?

rduivenvoorde commented:
Hi Cory,

NOTE: your are using QGIS 3.20 (or the development version) yes? There has been a changes in 3.20 in temporal code.

I'm pretty sure we should be able to 'fix' this using virtual fields. I've used it to create 'DateTimes' for a time Txt column).
But a lot of factors are in play here.

What is your original data source? shp files or postgis or ???
Is it possible for you to sent me ***@***.***) a small dataset including a project file?

Then I will have a look into it (and try to find a work around).

Regards,

Richard Duivenvoorde


On 8/10/21 8:56 PM, Cory Albrecht wrote:
> OK, then how do you recommend I work around this issue in the meantime?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/qgis/QGIS/pull/40989#issuecomment-896235727>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAFSUGNZ5QXYUCMUK37J6HDT4FY5VANCNFSM4WA3QLRA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
>

CoryAlbrecht commented:
@rduivenvoorde I currently install from `deb [arch=amd64] http://qgis.org/ubuntu-nightly-release focal main` (I am back on Ubuntu 20.04 LTS after a few months on Windows 10). However, I only update irregularly, once a week at best, some times not for a month.

Original data is PostGIS, and my start and end fields are both PostgreSQL `DATE` columns.

I added an example project over on the ticket I created. Did you need something different?
